src/event/AsyncEventTrait.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+/*
+ *
+ *  ____            _        _   __  __ _                  __  __ ____
+ * |  _ \ ___   ___| | _____| |_|  \/  (_)_ __   ___      |  \/  |  _ \
+ * | |_) / _ \ / __| |/ / _ \ __| |\/| | | '_ \ / _ \_____| |\/| | |_) |
+ * |  __/ (_) | (__|   <  __/ |_| |  | | | | | |  __/_____| |  | |  __/
+ * |_|   \___/ \___|_|\_\___|\__|_|  |_|_|_| |_|\___|     |_|  |_|_|
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * @author PocketMine Team
+ * @link http://www.pocketmine.net/
+ *
+ *
+ */
+
+declare(strict_types=1);
+
+namespace pocketmine\event;
+
+use pocketmine\promise\Promise;
+use pocketmine\promise\PromiseResolver;
+use pocketmine\utils\ObjectSet;
+use function array_shift;
+use function count;
+
+trait AsyncEventTrait {
+	/** @phpstan-var ObjectSet<Promise<null>> */
+	private ObjectSet $promises;
+
+	/**
+	 * @phpstan-param ObjectSet<Promise<null>>|null $promises
+	 */
+	private function initializePromises(?ObjectSet &$promises) : void{
+		$promises ??= new ObjectSet();
+		$this->promises = $promises;
+	}
+
+	public function addPromise(Promise $promise) : void{
```

Comment:
Is it an expected use case that a single handler adds multiple promises?

Reply:
Yes, this will be handled as if multiple handlers were adding promises. 

Added promises are only processed once all handlers of the priority have been called, so no difference will be made on this side.

Reply:
What if user calls `addPromise` on the event again after the initial dispatch? Your `Promise::all()` call won't include the new promises, and we cannot prevent them from doing so.

Reply:
We should throw an exception if `addPromise` is called in an asynchronous context, to avoid this confusion.